GameServer/gameobjects/Currency.cs

```diff
@@ -0,0 +1,108 @@
+using System.Collections.Generic;
+using DOL.Database;
+
+namespace DOL.GS.Finance
+{
+    public abstract class Currency
+    {
+        private enum eCurrency : byte
+        {
+            Copper = 1,
+            ItemTemplate = 2,
+            BountyPoints = 3,
+            Mithril = 4
+        }
+
+        private static Dictionary<eCurrency, Currency> cachedTypes = new Dictionary<eCurrency, Currency>(){
+            {eCurrency.Copper, new Copper()},
+            {eCurrency.BountyPoints, new BountyPoints()},
+            {eCurrency.Mithril, new Mithril()},
+        };
```

Comment:
How can I not take this as an unsubstatiated insult? Scattering flyweight factory logic across sub classes is definitely worse than what I did here. Removing eCurrency would make sense, however that evolved to that point over time, so I still don't accept the insult.

Reply:
I'm deeply sorry, it wasn't my intention at all, I think my english is too bad to understand exactly what I said, I don't really see where my insult is :(

In my head, the code was like that just with the evolution of your changes and should need a small iteration to remove some dependencies between classes.
What I wanted to say about the factory part, it's just to have a singleton-like for sub classes (except ItemCurrency which need an instance per item template) since it's nearly a constant but it can be a simple property in Currency (there is no need for the dictionary).

Reply:
Yeah I agree with the latter part, that a dictionary is not really necessary.  This is btw not meant to be a singleton, but only meant as an optimization measure (maybe premature, but it felt sensible). It should work with multiple instances of the same object as well.

Reply:
I'm not talking about optimisation (even if it can be) but more about semantic : Copper will be the same copper everywhere, same for BountyPoints or Mithril, it's why I'm saying to have only one instance for theses.